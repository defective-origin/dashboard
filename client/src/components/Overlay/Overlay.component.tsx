import React, { useRef } from 'react'

// ---| core |---
import { cn } from 'tools'
import { ElementOptions, useMode } from 'hooks'

// ---| pages |---
// ---| screens |---
// ---| components |---

// ---| self |---
import css from './Overlay.module.scss'

export type OverlaySideVariant = 'left' | 'right' | 'top' | 'bottom'
export type OverlaySideCernerVariant = 'left-center' | 'right-center' | 'top-center' | 'bottom-center'
export type OverlayCernerVariant = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right'
export type OverlayCenterVariant = 'center' | 'full'
export type OverlayVariant = OverlaySideVariant | OverlaySideCernerVariant | OverlayCernerVariant | OverlayCenterVariant

export type OverlayProps = React.HTMLAttributes<HTMLDivElement> & {
  v?: OverlayVariant
  width?: React.CSSProperties['width']
  height?: React.CSSProperties['height']
  indent?: React.CSSProperties['margin']
  window?: boolean
  backdrop?: boolean
  contentClassName?: string
  /** Set class name on overlay container when overlay mounted. */
  containerClassName?: string
  /** Container selector. If not passed then takes first parent node. */
  container?: ElementOptions<HTMLElement>
}

/**
 * Allows to create overlay's components like: Modal, Drawer, DIalog and so on.
 * Adds position relative to parent component if parent component has position static.
 *
 * How to use
 * @example
 * export class ModalBreakpoint implements Breakpoint {
 *   constructor(
 *     public name?: string,
 *     public size = Number.MAX_SAFE_INTEGER,
 *   ) {}
 * }
 *
 * export const Modal = (props: OverlayProps) => {
 *    const { v, children, ...otherProps } = props
 *    const breakpoints = useMemo(() => [
 *      new ModalBreakpoint('full', 768),
 *      new ModalBreakpoint(v),
 *    ], [v])
 *
 *    const breakpoint = useBreakpoint(breakpoints)
 *
 *    return (
 *      <Overlay ref={breakpoint.ref} v={breakpoint.name} indent={10} {...otherProps}>
 *        {children}
 *      </Overlay>
 *    )
 * }
 */
export function Overlay(props: OverlayProps): JSX.Element {
  const {
    v = 'center',
    width,
    height,
    indent,
    backdrop,
    window,
    container = () => overlayRef.current?.parentElement,
    className,
    contentClassName,
    containerClassName,
    children,
    ...otherProps
  } = props
  const overlayRef = useRef<HTMLDivElement>(null)
  const _className = cn(css.Overlay, {
    [css.Backdrop]: backdrop,
    [css.Window]: window,
  }, className)

  const _contentClassName = cn(css.OverlayContent, css[v], contentClassName)
  const styles = {
    margin: indent,
    width: v === 'full' ? undefined : width,
    height: v === 'full' ? undefined : height,
  }

  useMode(container, containerClassName, (element) => {
    const position = element && getComputedStyle(element).getPropertyValue('position')
    if (position && position === 'static' && !window) {
      return css.OverlayContainer
    }
  })

  return (
    <div className={_className} ref={overlayRef}>
      <div className={_contentClassName} style={styles} {...otherProps}>
        {children}
      </div>
    </div>
  )
}

Overlay.displayName = 'Overlay'

export default Overlay
