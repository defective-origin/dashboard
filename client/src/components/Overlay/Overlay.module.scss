@import 'src/theme';

:root {
  --popup-backdrop-color: var(--color--primary-9);
}

.OverlayContainer { position: relative; }

.Overlay { position: absolute; pointer-events: none; inset: 0; }
.Window { position: fixed; }

.Backdrop {
  background-color: var(--popup-backdrop-color);
  pointer-events: all;
}

.OverlayContent {
  display: grid;
  position: absolute;
  min-width: 0;
  min-height: 0;
  pointer-events: all;
}

// side variants
.left { top: 0; bottom: 0; left: 0; }
.right { top: 0; bottom: 0; right: 0; }
.top { right: 0; left: 0; top: 0; }
.bottom { right: 0; left: 0; bottom: 0; }

// cerner variants
.top-left { top: 0; left: 0; }
.top-right { top: 0; right: 0; }
.bottom-left { bottom: 0; left: 0; }
.bottom-right { bottom: 0; right: 0; }

// side-center variants
.left-center {  top: 50%; left: 0; transform: translateY(-50%); }
.right-center { top: 50%; right: 0; transform: translateY(-50%); }
.top-center { left: 50%; top: 0; transform: translateX(-50%); }
.bottom-center { left: 50%; bottom: 0; transform: translateX(-50%); }

// center variants
.full { inset: 0; }
.center { top: 50%; left: 50%; transform: translate(-50%, -50%); }
