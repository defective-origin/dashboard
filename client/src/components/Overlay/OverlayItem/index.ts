/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './OverlayItem.component'

/* INJECT_EXPORT_PLACE */
export * from './OverlayItem.component'
