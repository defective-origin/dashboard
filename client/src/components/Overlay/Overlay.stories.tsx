// eslint-disable-next-line no-restricted-imports
import { SB_CSS, field, params } from '../../../.storybook/tool'
import type { Meta, StoryObj } from '@storybook/react'
import Item from 'components/Item'
import Block from 'components/Block'
import Overlay, { OverlayProps } from './Overlay.component'

const SIDE_VARIANTS: OverlayProps['v'][] = ['left', 'right', 'top', 'bottom']
const SIDE_CERNER_VARIANTS: OverlayProps['v'][] = ['left-center', 'right-center', 'top-center', 'bottom-center']
const CERNER_VARIANTS: OverlayProps['v'][] = ['top-left', 'top-right', 'bottom-left', 'bottom-right']
const CENTER_VARIANTS: OverlayProps['v'][] = ['center', 'full']
const VARIANTS: OverlayProps['v'][] = [...SIDE_VARIANTS, ...SIDE_CERNER_VARIANTS, ...CERNER_VARIANTS, ...CENTER_VARIANTS]

const meta: Meta<typeof Overlay> = {
  title: 'Components/LAYOUT/Overlay',
  // throws ref error if we pass as 'component: Overlay'
  component: (props) => <Overlay {...props} />,
  tags: ['autodocs'],
  argTypes: {
    className: field.string(),
    contentClassName: field.string(),
    containerClassName: field.string(),
    children: field.reactNode(),
    v: field.variants(VARIANTS, 'OverlayVariant'),
    window: field.boolean(),
    backdrop: field.boolean(),
    width: field.number('Width'),
    height: field.number('Height'),
    indent: field.number('Margin'),
    container: field.element(),
  },
}

export default meta

type Story = StoryObj<typeof Overlay>

const render = (props: OverlayProps) => (
  <Block width={400} height={400} border={SB_CSS.border}>
    <Overlay {...props}>
      <Item stretch background={SB_CSS.item} minWidth={50} minHeight={50} />
    </Overlay>
  </Block>
)

export const Demo: Story = {
  parameters: params('Overlay'),
  render,
  args: {
    v: 'center',
    window: false,
    backdrop: false,
  },
}

export const Variants: Story = {
  parameters: params('View', VARIANTS),
  render,
  args: {
    v: 'center',
  },
}

export const Position: Story = {
  parameters: params('Position'),
  render,
  args: {
    v: 'center',
    window: true,
  },
}

export const Backdrop: Story = {
  parameters: params('Backdrop'),
  render,
  args: {
    v: 'center',
    backdrop: true,
  },
}

export const Size: Story = {
  parameters: params('Size'),
  render,
  args: {
    v: 'center',
    width: 300,
    height: 300,
  },
}

export const Indent: Story = {
  parameters: params('Indent'),
  render,
  args: {
    v: 'center',
    indent: 10,
  },
}
