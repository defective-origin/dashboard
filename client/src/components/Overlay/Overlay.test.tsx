// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Overlay from './Overlay.component'
// import useOverlay from './Overlay.hook'
// import OverlayProvider, Overlay from './Overlay.context'

describe('[Overlay] component', () => {
  it('should render component', () => {
    const container = render(<Overlay />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})

// describe('[Overlay] hook', () => {
//   it('should return options', () => {
//     const { result } = renderHook(() => useOverlay())
//
//     expect(result.current).toEqual(null)
//   })
// })

// describe('[Overlay] context', () => {
//   const wrapper = (props: OverlayProps) => <OverlayProvider { ...props } />
//
//   it('should return options', () => {
//     const { result } = renderHook(() => useOverlay(), { wrapper })
//
//     expect(result.current).toBeTruthy()
//   })
// })
