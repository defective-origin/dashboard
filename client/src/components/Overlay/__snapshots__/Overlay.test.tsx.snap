// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Overlay] component > should render component 1`] = `
<DocumentFragment>
  <div
    class="_Overlay_dfe0de"
  >
    <div
      class="_OverlayContent_dfe0de _center_dfe0de"
    />
  </div>
</DocumentFragment>
`;
